### PR TITLE
Chapter4

### PR DESCRIPTION
**Pull request by carsen-stringer**
we did not add anything to the glossary, happy to help with this once all the chapters are in

### CODE REVIEW

**Comment by carsen-stringer** [comment]
fyi @ScientistRachel added the outline of chapter 5 on this branch (I did not edit Chapter 5)

File: docs/4-architectures/same_padding_no_strides.gif
Comment (opp1231) [inline docs/4-architectures/same_padding_no_strides.gif]: See previous comment on GIF's.

File: docs/4-architectures.qmd [hunk 1]
```diff
@@ -7,129 +7,351 @@ author:
...
+
+## Neural network architectures
+
+In computer vision, multiple neural network architectures were designed for various visual tasks. These architectures take as input an image, which often has multiple channels - in the case of natural images there are three input channels for red/green/blue. The neural networks process the image and output various quantities depending on the task. For example, in the case of object recognition, the networks output a single probability vector to indicate the likelihood that each possible object (cat, dog etc) is in the image. 
 
-Starting prompt for this chapter: Chapter 4 introduces architectures and loss models, defining them and providing examples through two practical case studies: image restoration and segmentation. Although this chapter will include code snippets/exercises, the presentation of essential concepts should communicate the philosophy behind the choice of a model for non-programmers.
+![Alexnet architecture, adapted from [@krizhevsky2012imagenet]. This early neural network first demonstrated the capabilities of deep learning when trained on large datasets.](4-architectures/alexnet.PNG){#fig-alexnet width=90%}
```
Comment (opp1231) [inline docs/4-architectures.qmd]: Re: Figures 1 and 2. For those people who have never seen any architecture diagram before, they may not grasp the meaning of these figures. Adding a sentence or two of context for those individuals would be helpful. Alternatively, would you prefer to move these figures deeper into the chapter where the concepts are explained more in-depth?

File: docs/4-architectures.qmd [hunk 2]
```diff
@@ -7,129 +7,351 @@ author:
...
+
+Taking our example from above, let’s estimate the number of parameters with filters/kernels of size 3 by 9 by 9 pixels, where 3 is the number of input channels and 9 is the size in pixel space, the “kernel size”. The number of input and output images are called “channels”, similar to the red/green/blue channels for RGB images. If we define the layer to have 6 output channels, this requires 6 of these kernels resulting in 1458 parameters in the kernels, along with 6 bias terms, resulting in 1464 parameters in total. As you can see, the number of parameters now is independent of the size of the input in pixels, and a dramatic reduction from the nearly 1 billion parameters for the dense linear layer example above.
```
Comment (opp1231) [inline docs/4-architectures.qmd]: The explanation of "channels" would be especially helpful if it were included earlier in the text.

Comment (opp1231) [inline docs/4-architectures.qmd]: Please break this into intermediate steps.

File: docs/4-architectures.qmd [hunk 3]
```diff
@@ -7,129 +7,351 @@ author:
...
+with torch.no_grad():
+    y = layer(X)
+
+print('input shape: ', X.shape)
+print('output shape: ', y.shape)
 ```
 
+The activations of a single convolutional layer, as described, have a receptive field size equivalent to its kernel size: each activation only receives information from pixels within the kernel size. However, objects in images are often larger than the kernel size. To increase the amount of spatial information used for computation, pooling layers are introduced between convolutional layers. A pooling layer consists of a sliding-window operation, just like the convolution layer, but in this case a maximum or average operation is computed within the window, independently for each input channel. To reduce the size of the image, this convolutional operation is applied with a “stride”: to downsample the image by a factor of two, we may use a pooling size of 2 with a stride also set to 2, like in the example below. 
+
+![Illustration of max-pooling with a kernel size of 2 and stride of 2, from [here](https://github.com/dvgodoy/PyTorchStepByStep).](4-architectures/pooling.png){#fig-pool width=90%}
```
Comment (opp1231) [inline docs/4-architectures.qmd]:
```suggestion
![Illustration of max-pooling with a kernel size of 2 and stride of 2 on an image with one channel, one z-slice and 6 pixels in x and y. The output only has 3 pixels in x and y due to the pooling step. Example from from [here](https://github.com/dvgodoy/PyTorchStepByStep).](4-architectures/pooling.png){#fig-pool width=90%}
```

File: docs/4-architectures.qmd [hunk 4]
```diff
@@ -7,129 +7,351 @@ author:
...
+# max pooling operation
+pool_layer = nn.MaxPool2d(kernel_size=2, stride=2)
+y = pool_layer(X)
+
+print('input shape: ', X.shape)
+print('output shape: ', y.shape)
 ```
 
-Here is an example equation.
+If we apply several convolutional and pooling layers, we end up with an output which is smaller in resolution than the input. We can see example activations across convolutional, pooling and linear layers in a network trained to classify MNIST [@lecun1998mnist] digits:
```
Comment (opp1231) [inline docs/4-architectures.qmd]:
```suggestion
If we apply several convolutional and pooling layers, we end up with an output which is smaller in resolution than the input. We can see example activations across convolutional, pooling and linear layers in a network trained to classify MNIST [@lecun1998mnist] digits (a dataset of handwritten numbers often used to test ML ideas):
```

File: docs/4-architectures.qmd [hunk 5]
```diff
@@ -7,129 +7,351 @@ author:
...
 
-$$
-s = \sqrt{\frac{1}{N-1} \sum_{i=1}^N (x_i - \overline{x})^2}
-$$ {#eq-stddev}
+![Activations across convolutional and pooling layers from this [demo](https://adamharley.com/nn_vis/cnn/2d.html). Lighter blue indicates a higher activation.](4-architectures/mnist.PNG){#fig-mnist}
```
Comment (opp1231) [inline docs/4-architectures.qmd]:
```suggestion
![Each layer of a small, illustrative network where the output channels of each layer are shown as images. Lighter blue indicates higher activation at each layer. The final output layer represents the probability of the input image corresponding to a value between 0 and 9. In this example, the network correctly identifies the input as the number 3. To see this in action, checkout the interactive demonstration [here](https://adamharley.com/nn_vis/cnn/2d.html).](4-architectures/mnist.PNG){#fig-mnist}
```

File: docs/4-architectures.qmd [hunk 6]
```diff
@@ -7,129 +7,351 @@ author:
...
+### Vision transformers
 
-## Quarto has additional features.
+Vision transformers are modern architectures that are replacing convolutional networks in many applications. They are not as parameter-efficient as convolutional neural networks - for example the Cellpose segmentation u-net has 6.6 million parameters while ViT-H (“vision-transformer-huge”) has 632 million parameters [@dosovitskiy2020image]. They are still much more efficient than dense linear layers due to special architecture choices (see below), and they introduce a new type of operation called “self-attention”. Transformers avoid overfitting this large set of parameters through training on very large amounts of data. Even though they have many more parameters than standard convolutional networks, they are not too much slower because most of the operations within the transformer are matrix multiplications which are fast on newer GPUs with tensor cores (e.g. from [nvidia](https://www.nvidia.com/en-us/data-center/tensor-cores/)). With more parameters, they have a larger capacity than standard convolutional neural networks to learn from large training datasets.
```
Comment (opp1231) [inline docs/4-architectures.qmd]: This is potentially silly, but it is possible the Janelia lawyers will not appreciate linking to a specific product. Is it possible to refer to "tensor cores" without linking to NVIDIA directly?

File: docs/4-architectures.qmd [hunk 7]
```diff
@@ -7,129 +7,351 @@ author:
...
+
+## Loss functions
+
+In a standard image classification network, the output is a vector with  length equal to the number of classes in the task. Each of the entries in this vector represents the predicted probability of the class, and the predicted label for the image is chosen as the index of the vector with the largest entry. For each training image we have a ground-truth label for the class. How close the network matches the label is the loss, which is defined as a function between the vector output of the network and the ground-truth label. A lower loss means we matched the ground-truth data better. The gradients of the network are computed automatically via back-propagation, and an optimizer is specified to modify the parameters in order to minimize the loss (described more below).
```
Comment (opp1231) [inline docs/4-architectures.qmd]: This is great to include!

File: docs/4-architectures.qmd [hunk 8]
```diff
@@ -7,129 +7,351 @@ author:
...
+
+$$ \ell(x, y) = - \sum_{c=1}^C \log [p_c(x)] y_c.$$
+
+In segmentation and biological classification tasks, as mentioned before, the output is often the same size as the input in pixels and the loss is computed per-pixel. There will thus be multiple outputs of this size, each one corresponding to a class, like the entries in the vector for overall image classification. In the original u-net paper, the loss was defined using two classes, "not cell" and "cell" [@ronneberger2015u]. 
+
+![Binary cross-entropy loss, computed from the output of a u-net trained to classify cell/not cell (predicted cell probability).](4-architectures/bceloss.PNG){#fig-bce width=70%}
```
Comment (opp1231) [inline docs/4-architectures.qmd]: A walk-through of this figure in the text would add a lot of context for it.

File: docs/4-architectures.qmd [hunk 9]
```diff
@@ -7,129 +7,351 @@ author:
...
+loss = loss_fn(y, y_true)
+
+print('loss = ', loss.item())
 ```
-:::
 
-::: {.callout-warning}
-Be careful to avoid hallucinations.
-:::
+To create segmentations for each cell, a threshold is defined on the cell probability and any pixels above the threshold that are connected to each other are formed into objects. This threshold is defined using a validation set - images that are not used for training or testing - to help ensure the threshold generalizes to the held-out test images. The predicted segmentations with this loss function often contain several merges, because cells can often touch each other and the connected components of the image will combine the touching cells into single components.
```
Comment (opp1231) [inline docs/4-architectures.qmd]: these concepts are first explained in Section 4.3. Perhaps consider moving that explanation up or cross-reference that section here.

File: docs/4-architectures.qmd [hunk 10]
```diff
@@ -7,129 +7,351 @@ author:
...
+
+## Training neural networks
+
+Now that we have defined a loss function, we want to minimize the loss $\ell$ by updating the weights in the network. For a problem like segmentation, we will need images with ground-truth segmentation labels - this labeling can be done using tools like [Ilastik](https://www.ilastik.org/), [ImageJ](https://imagej.net/software/imagej/), [Paintera](https://github.com/saalfeldlab/paintera), or [Napari](https://napari.org/stable/). Once ground-truth labeling is done on some images, training can be attempted. We will want to use most of the ground-truth labeled images for training, making up a training set, and leave a small subset (like 10%) for testing the performance of the network. For some algorithms, we may also need a validation set for setting post-processing parameters like the cell probability threshold, in which case we can reserve 10-15% of the training set images for validation.
+
+Minimizing the loss is done via gradient descent, in which the weights are updated by the gradient of the loss with respect to each parameter, scaled by the learning rate $\alpha$:
```
Comment (opp1231) [inline docs/4-architectures.qmd]: Please break this down into intermediate steps.

File: docs/4-architectures.qmd [hunk 11]
```diff
@@ -7,129 +7,351 @@ author:
...
+
+During fitting it is important to monitor the training loss and the validation loss. With an appropriate learning rate that is not too large, the training loss should always decrease. The loss on held-out validation data should also ideally decrease over training. If not, then the network is overfitting to the training set: the weights are becoming specifically tuned for the training set examples and no longer generalize to held-out data.
```
Comment (opp1231) [inline docs/4-architectures.qmd]: This is great to mention.

File: docs/4-architectures.qmd [hunk 12]
```diff
@@ -7,129 +7,351 @@ author:
...
+
+![Example training loss and validation loss across epochs.](4-architectures/trainloss.png){#fig-trainloss width=40%}
+
+To avoid overfitting, regularization is often used. Most commonly in computer vision problems, weight decay will be used for regularization, which is closely-related to L2 regularization. This operation reduces the weights by a small fraction $\lambda$ at each optimization step:
```
Comment (opp1231) [inline docs/4-architectures.qmd]: We don't think the intended audience is likely familiar with regularization at all. We think this is a good change to generalize this explanation.

File: docs/4-architectures.qmd [hunk 13]
```diff
@@ -7,129 +7,351 @@ author:
...
+ 
+Moving the parameters in the negative direction of the gradient reduces the loss for the given images or data points over which the loss is computed. We could compute the loss and gradients over all images in the training set, but this would take too long so in practice the loss is computed in batches of a few to a few hundred images - the number of images in a batch is called the *batch size*. The optimization algorithm for updating the weights in batches is called stochastic gradient descent (SGD). This is often faster than full-dataset gradient descent because it updates the parameters many times on a single pass through the training set (called “epoch”). Also, the stochasticity induced by the random sampling step in SGD effectively adds some noise in the search for a good minimum of the loss function, which may be useful for avoiding local minima.
+
+It can also be beneficial to include momentum, with some value $\beta$ between zero and one, which pushes weight updates along the same direction they have been updating in the past. The updated version of $\vec{v}$ in this case is
```
[the remark this answers is not in the document]
Comment (carsen-stringer) [reply docs/4-architectures.qmd]: FYI this is the classical reason for momentum https://en.wikipedia.org/wiki/Gradient_descent#Momentum_or_heavy_ball_method

File: docs/4-architectures.qmd (hunk 9 above)
Comment (carsen-stringer) [reply docs/4-architectures.qmd]: Sorry I'm not sure if there is a repeat here, but after I push my changes, please make another suggestion

**Comment by carsen-stringer** [comment]
Hi @opp1231 and @ScientistRachel I pushed my updates, addressing all of @opp1231's comments (except for the repetition question), and I created two new figures to try to clarify the linear layer section. Please let me know if anything is still unclear.
